prometheus/registry.go

```diff
@@ -368,22 +376,12 @@ func (r *Registry) Gather() ([]*dto.MetricFamily, error) {
 	)
 
 	r.mtx.RLock()
+	goroutineBudget := len(r.collectorsByID)
```

Comment:
is there a reasonable maximum that we could put here to limit memory usage? Something like "never start more than 100k goroutines" (~ 40MB) so that in the pathological case of many slow collectors, the collection time suffers but we don't overload the system?

Reply:
I don't think there is a one-size-fits-it-all maximum. If you have a large system, 1M goroutines might be what you want. If you are on a small system, you want fewer, but if you have 1M slow collectors on a small system, you are doomed anyway.

Reply:
Or in other words: I find it reasonable to start one goroutine per slow collector. And that's what this PR accomplishes in an adaptive fashion.